README.rst
================================================================================
pyramid_assetcompiler v0.1
================================================================================

| Copyright: (c) 2012 Seth Davis
| http://github.com/seedifferently/pyramid_assetcompiler


Synopsis
================================================================================

pyramid_assetcompiler provides automatic asset compiling for your Pyramid
applications.


Installation
================================================================================

To install, simply::

    pip install pyramid_assetcompiler

* You'll need to have `Python`_ 2.6+ and `pip`_ installed.

.. _Python: http://www.python.org
.. _pip: http://www.pip-installer.org


Documentation
================================================================================

You can read the documentation at: http://pyramid_assetcompiler.readthedocs.org/


Known Issues and Limitations
================================================================================

* Hasn't yet been tested against Python 3 (initial versions are working concepts
  only).
* Hasn't yet been tested against Unicode filenames (see above).
* Kind of obvious, but you have to actually have the specified compiler command
  installed and accessible from your working path in order for it to work.


Disclaimers and Warnings
================================================================================

This is Alpha software--use at your own risk!

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHOR BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGES.rst
Latest Changes
================================================================================


v0.1 -- 10/21/2012
--------------------------------------------------------------------------------

* Initial release.


pyramid_assetcompiler/compiler.py
# vim:fileencoding=utf-8:ai:ts=4:sts:et:sw=4:tw=80:
import os, re, glob, shlex, subprocess, hashlib
from pyramid.path import AssetResolver

class Compiler(object):
    """
    Compiler class for the pyramid_assetcompiler add-on.
    """
    def __init__(self, app_settings, path, **kw):
        """
        Initialize the Compiler class.
        
        Required parameters:
        
        :type app_settings: dict
        :param app_settings: The Pyramid application ``settings`` dictionary.
        
        :type path: string
        :param path: The Pyramid ``asset path``.
        
        Optional parameters:
        
        :type compiler: dict or string
        :param compiler: Allows you to either specify a specific compiler to
                         use (e.g. ``coffee``), or assign a brand new
                         compiler dictionary to be used (e.g.
                         ``{'less': {'cmd': 'lessc', 'ext': 'css'}}``)
        
        :type batch: bool
        :param batch: Specify that the class should prepare for a batch
                      compile rather than a normal compile.
        """
        self.settings = app_settings
        self.path = path
        
        self.compilers = self.settings.get('assetcompiler.compilers')
        self.prefix = self.settings.get('assetcompiler.asset_prefix')
        
        if not isinstance(self.compilers, dict):
            raise ValueError('No compilers were found.')
        
        self.compiler = kw.get('compiler', None)
        self.batch = kw.get('batch', False)
        self.hash = None
        self.exists = None
        
        self.fullpath = self.path
        if not os.path.isabs(self.path):
            # Try to resolve the asset full path
            a = AssetResolver()
            self.fullpath = a.resolve(self.path).abspath()
        
        if self.batch:
            if not os.path.isdir(self.fullpath):
                raise EnvironmentError('Directory does not exist: %s' % \
                                       self.fullpath)
        
        else:
            self.filename = os.path.basename(self.fullpath)
            self.dirname = os.path.dirname(self.fullpath)
            self.name = os.path.splitext(self.filename)[0]
            self.ext = os.path.splitext(self.filename)[1][1:]
            
            if self.compiler:
                if not isinstance(self.compiler, dict):
                    self.compiler = self.compilers.get(self.compiler, {})
            else:
                self.compiler = self.compilers.get(self.ext, {})
            
            
            if not self.compiler.get('cmd') or not self.compiler.get('ext'):
                raise ValueError('No compiler found for %s' % self.ext)
    
    
    @property
    def compiled(self):
        """
        Property method to check and see if the initialized asset path has
        already been compiled.
        """
        new_ext = self.compiler['ext']
        
        self.hash = self.hash or self._compute_hash(self.fullpath)
        self.new_filename = '%s%s.%s.%s' % (self.prefix, self.name, self.hash,
                                            new_ext)
        self.new_fullpath = os.path.join(self.dirname, self.new_filename)
        self.new_path = re.sub(r'%s$' % self.filename, self.new_filename,
                                        self.path)
        
        self.exists = self.exists or self._check_exists(self.new_fullpath)
        
        return self.exists
    
    def _check_exists(self, path):
        """
        Convenience method to check if a file already exists.
        """
        if os.path.exists(path):
            return True
        else:
            return False
    
    def _compute_hash(self, path):
        """
        Convenience method to compute the hash key for the compiled asset.
        """
        md5 = hashlib.md5()
        
        # Loop the file, adding chunks to the MD5 generator
        with open(path, 'rb') as f: 
            for chunk in iter(lambda: f.read(128*md5.block_size), b''): 
                md5.update(chunk)
        # Finally, add the mtime
        md5.update(str(os.path.getmtime(path)))
        
        # Get the first 12 characters of the hexdigest
        self.hash = md5.hexdigest()[:12]
        
        return self.hash
    
    def compile(self):
        """
        Runs the compiler for the initialized asset.
        """
        cmd = '%s %s' % (self.compiler['cmd'], self.fullpath)
        
        proc = subprocess.Popen(
            shlex.split(cmd),
            stdout=subprocess.PIPE,
            stdin=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
        out, err = proc.communicate()
        
        if proc.returncode != 0 or err:
            raise EnvironmentError('%s\n\n%s' % (err, out))
        else:
            new_dirname = os.path.normpath(os.path.dirname(self.new_fullpath))
            
            if not os.path.exists(new_dirname):
                os.makedirs(new_dirname)
            
            with open(self.new_fullpath, 'w') as f:
                f.write(out)
            
            return self.new_path
    
    def batch_compile(self):
        """
        Runs the compiler for the initialized batch of assets.
        """
        for ext, data in self.compilers.items():
            for asset in glob.glob(os.path.join(self.fullpath, '*.%s' % ext)):
                filename = os.path.basename(asset)
                dirname = os.path.dirname(asset)
                name = os.path.splitext(filename)[0]
                new_ext = data['ext']
                
                self.hash = self.hash or self._compute_hash(asset)
                new_filename = '%s%s.%s.%s' % (self.prefix, name, self.hash,
                                               new_ext)
                new_fullpath = os.path.join(dirname, new_filename)
                
                if not os.path.exists(new_fullpath):
                    cmd = '%s %s' % (data['cmd'], asset)
                    
                    proc = subprocess.Popen(
                        shlex.split(cmd),
                        stdout=subprocess.PIPE,
                        stdin=subprocess.PIPE,
                        stderr=subprocess.PIPE,
                    )
                    out, err = proc.communicate()
                    
                    if proc.returncode != 0 or err:
                        raise EnvironmentError('%s\n\n%s' % (err, out))
                    else:
                        new_dirname = os.path.normpath(
                            os.path.dirname(new_fullpath)
                        )
                        
                        if not os.path.exists(new_dirname):
                            os.makedirs(new_dirname)
                        
                        with open(new_fullpath, 'w') as f:
                            f.write(out)
    
    def compiled_data(self):
        if not self.exists:
            raise ValueError('Source not found. Has it been compiled?')
        
        with open(self.new_fullpath) as f:
            data = f.read()
        
        return data


pyramid_assetcompiler/__init__.py
# vim:fileencoding=utf-8:ai:ts=4:sts:et:sw=4:tw=80:
try:
    from collections import OrderedDict
except ImportError:
    # Py 2.6 compat
    from ordereddict import OrderedDict

from pyramid.settings import asbool
from pyramid.events import ApplicationCreated, BeforeRender
from pyramid.threadlocal import get_current_request

from pyramid_assetcompiler.utils import as_string, as_list
from pyramid_assetcompiler.compiler import Compiler

SETTINGS_PREFIX = 'assetcompiler.'

default_settings = (
    ('debug', asbool, 'false'),
    ('asset_prefix', as_string, '_'),
    ('each_request', asbool, 'true'),
    ('each_boot', asbool, 'false'),
    ('asset_paths', as_list, ('',)),
)

# Use an OrderedDict so that processing always happens in order
compilers = OrderedDict() # empty for now

def parse_settings(settings):
    parsed = {}
    def populate(name, convert, default):
        name = '%s%s' % (SETTINGS_PREFIX, name)
        value = convert(settings.get(name, default))
        parsed[name] = value
    for name, convert, default in default_settings:
        populate(name, convert, default)
    return parsed

def assign_compiler(config, ext, cmd, new_ext):
    """
    Configuration method to set up an asset compiler and its parameters.
    
    :param ext: The file extension to match (e.g. coffee).
    :type ext: string - Required
    
    :param cmd: The command to run (e.g. coffee -c -p). The filename to be
                compiled will automatically be appended to the end of this
                string.
    :type cmd: string - Required
    
    :param new_ext: The extension that the compiled file should have (e.g.
                    js).
    :type new_ext: string - Required
    
    
    .. warning:: The specified compiler command must be installed, must be
                 executable by the Pyramid process, and must *output the
                 compiled data to stdout*. The last point can get tricky
                 depending on the command, so be sure to check its command
                 switches for the appropriate option.
    
    Examples:
    
    A compiler that will compile CoffeeScript files into JavaScript::
    
        config.assign_compiler('coffee', 'coffee -c -p', 'js')
    
    You can even nest compilers. This could allow you to compile CoffeeScript
    files into JavaScript, and then minify the JavaScript (see the docs for
    ``compiled_assetpath``)::
    
        config.assign_compiler('coffee', 'coffee -c -p', 'js')
        config.assign_compiler('js', 'uglifyjs', 'js')
    
    Note: Compilers are first in first out, so make sure you assign accordingly.
    """
    compilers[ext] = dict(cmd=cmd, ext=new_ext)

def compiled_asset_url(path, **kw):
    """
    Returns a Pyramid ``static_url`` url of the compiled asset (and compiles the
    asset if needed).
    
    :param path: The Pyramid asset path to process.
    :type path: string - Required
    
    :type compiler: dict or string - Optional
    :param compiler: Allows you to override/specify a specific compiler to use
                     (e.g. ``coffee``), or assign a brand new compiler
                     dictionary to be used (e.g. ``{'less': {'cmd': 'less',
                     'ext': 'css'}}``)
    """
    request = get_current_request()
    settings = request.registry.settings
    
    compiler = Compiler(settings, path, **kw)
    
    if not settings.get('assetcompiler.each_request'):
        if not compiler.compiled:
            # log an error
            pass
        
        return request.static_url(compiler.new_path)
    else:
        if compiler.compiled:
            return request.static_url(compiler.new_path)
        else:
            return request.static_url(compiler.compile())


def compiled_asset_path(path, **kw):
    """
    Returns a Pyramid ``static_path`` path of the compiled asset (and compiles the
    asset if needed).
    
    :param path: The Pyramid asset path to process.
    :type path: string - Required
    
    :type compiler: dict or string - Optional
    :param compiler: Allows you to override/specify a specific compiler to use
                     (e.g. ``coffee``), or assign a brand new compiler
                     dictionary to be used (e.g. ``{'less': {'cmd': 'less',
                     'ext': 'css'}}``)
    """
    request = get_current_request()
    settings = request.registry.settings
    
    compiler = Compiler(settings, path, **kw)
    
    if not settings.get('assetcompiler.each_request'):
        if not compiler.compiled:
            # log an error
            pass
        
        return request.static_path(compiler.new_path)
    else:
        if compiler.compiled:
            return request.static_path(compiler.new_path)
        else:
            return request.static_path(compiler.compile())

def compiled_asset_source(path, **kw):
    """
    Returns the source data/contents of the compiled asset (and compiles the
    asset if needed). This is useful when you want to output inline data (e.g.
    for inline javascript blocks).
    
    :param path: The Pyramid asset path to process.
    :type path: string - Required
    
    :type compiler: dict or string - Optional
    :param compiler: Allows you to override/specify a specific compiler to use
                     (e.g. ``coffee``), or assign a brand new compiler
                     dictionary to be used (e.g. ``{'less': {'cmd': 'less',
                     'ext': 'css'}}``)
    """
    request = get_current_request()
    settings = request.registry.settings
    
    compiler = Compiler(settings, path, **kw)
    
    if not settings.get('assetcompiler.each_request'):
        if not compiler.compiled:
            # log an error
            return None
        
        return compiler.compiled_data()
    else:
        if compiler.compiled:
            return compiler.compiled_data()
        else:
            compiler.compile()
            return compiler.compiled_data()

def compiled_assetpath(path, **kw):
    """
    Returns a Pyramid ``asset`` path such as ``pkg:static/path/to/file.ext``
    (and compiles the asset if needed).
    
    :param path: The Pyramid asset path to process.
    :type path: string - Required
    
    :type compiler: dict or string - Optional
    :param compiler: Allows you to override/specify a specific compiler to use
                     (e.g. ``coffee``), or assign a brand new compiler
                     dictionary to be used (e.g. ``{'less': {'cmd': 'less',
                     'ext': 'css'}}``)
    
    This function could be used to nest ``pyramid_assetcompiler`` calls (e.g.
    ``compiled_asset_path(compiled_assetpath('pkg:static/js/script.coffee'))``
    could compile a CoffeeScript file into JS, and then further minify the JS
    file) with a compiler configuration such as::
    
        config.assign_compiler('coffee', 'coffee -c -p', 'js')
        config.assign_compiler('js', 'uglifyjs', 'js')
    """
    request = get_current_request()
    settings = request.registry.settings
    
    compiler = Compiler(settings, path, **kw)
    
    if not settings.get('assetcompiler.each_request'):
        if not compiler.compiled:
            # log an error
            pass
        
        return compiler.new_path
    else:
        if compiler.compiled:
            return compiler.new_path
        else:
            return compiler.compile()

def applicationcreated_subscriber(event):
    settings = event.app.registry.settings
    settings['assetcompiler.compilers'] = compilers
    
    if settings.get('assetcompiler.each_boot'):
        asset_paths = settings.get('assetcompiler.asset_paths')
        
        for asset_path in asset_paths:
            compiler = Compiler(settings, asset_path, batch=True)
            compiler.batch_compile()

def beforerender_subscriber(event):
    event['compiled_asset_url'] = compiled_asset_url
    event['compiled_asset_path'] = compiled_asset_path
    event['compiled_asset_source'] = compiled_asset_source
    event['compiled_assetpath'] = compiled_assetpath

def includeme(config):
    """
    Activate the package; usually called via
    ``config.include('pyramid_assetcompiler')`` instead of being invoked
    directly.
    """
    settings = parse_settings(config.registry.settings)
    config.registry.settings.update(settings)
    
    config.add_directive('assign_compiler', assign_compiler)
    config.add_subscriber(applicationcreated_subscriber, ApplicationCreated)
    config.add_subscriber(beforerender_subscriber, BeforeRender)
